refactor(major): file organization, build process

sarbbottam opened the pull request:
I guess, we both have been working in parallel for the same goal, ~~and thus there are merge conflicts~~.

~~Let's not worry about that right now.~~ I would request you to review the code. Before you start reviewing, allow me to brief you about the changes.

## Overview of changes.

* Refactor src/inject.js, please refer src/index.js
* use webpack instead of gulp
  * webpack will take care of the combining multiple js files as well as the css
  * no need to copy the css files
* chrome/src/inject.js, chrome/src/inject.js.map, firefox/data/inject.js & firefox/data/inject.js.map will be added to .gitignore, before merging the PR.

sarbbottam commented on chrome/manifest.json:
```diff
@@ -24,12 +24,9 @@
-      ],
       "js": [
-        "src/inject/inject.js"
+        "src/inject.js"
```
`inject` directory is not required

sarbbottam commented on chrome/manifest.json:
```diff
@@ -24,12 +24,9 @@
       "all_frames": false,
       "run_at": "document_end",
-      "css": [
-        "src/inject/inject.css"
```
`inject.css` is part of `inject.js`, see `src/index.js`

sarbbottam commented on chrome/src/inject/inject.css:
```diff
@@ -1,30 +0,0 @@
-.contrib-column {
```
`inject.css` is part of `inject.js`, see `src/index.js`

sarbbottam commented on chrome/src/inject/inject.js:
```diff
@@ -1,159 +0,0 @@
-"use strict";
```
`inject` directory is not required, see `chrome/src/inject.js`

eventually `chrome/src/inject.js` will be `.gitignore`d

sarbbottam commented on firefox/data/inject.css:
```diff
@@ -1,30 +0,0 @@
-.contrib-column {
```
`inject.css` is part of `inject.js`, see `data/index.js`

sarbbottam commented on firefox/index.js:
```diff
@@ -1,9 +1,8 @@
     contentScriptFile: self.data.url("inject.js"),
-    contentScriptWhen: "ready",
-    contentStyleFile: self.data.url("inject.css")
-});
```
`inject.css` is part of `inject.js`, see `data/index.js`

sarbbottam commented on package.json:
```diff
@@ -4,13 +4,10 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "eslint .; exit 0",
-    "install": "npm i -g babel-cli watch jpm",
```
`jpm` is part of `devDependencies`

sarbbottam commented on package.json:
```diff
@@ -4,13 +4,10 @@
-    "build:css": "echo ./firefox/data/ ./chrome/src/inject/ | xargs -n 1 cp ./src/inject.css",
-    "build": "npm run build:js && npm run build:css",
-    "build:watch": "watch \"npm run build\" ./src/",
+    "lint": "eslint .",
```
no `exit 0`; must fail if there are lint errors

sarbbottam commented on package.json:
```diff
@@ -4,13 +4,10 @@
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "eslint .; exit 0",
-    "install": "npm i -g babel-cli watch jpm",
-    "clean": "find build -regextype posix-egrep -regex \".*\\.(zip|xpi|nex)$\" -type f -delete && rmdir build",
```
see `prebuild`

sarbbottam commented on src/index.js:
```diff
@@ -0,0 +1,30 @@
+require('./index.css')
```
`index.js` is the main file, consider it as entry point.

sarbbottam commented on src/index.js:
```diff
@@ -0,0 +1,30 @@
+            inject()
+        }
+    })
+})
```
mutation observer logic is now simple, it aways observes `js-pjax-loader-bar` and invokes `inject` if desired.

sarbbottam commented on src/inject.js:
```diff
@@ -121,50 +119,4 @@ function inject() {
-            observer.observe(pixelBar, config)
-        }, false)
-    }))
-}
```
see `src/index.js`, this logic has been simplified.

sarbbottam commented on src/inject.js:
```diff
@@ -121,50 +119,4 @@ function inject() {
-            updateStyle(`li[style='background-color: #${key}']`, "backgroundColor", `#${colorsHash[key]}`)
-        })
-    }
-}
```
see `src/change-color.js`

Naramsim commented:
Well, these are a lot of big and great changes! :smile: 
Thanks for your effort @sarbbottam. Tomorrow I will take a deeper look.

sarbbottam commented:
> Well, these are a lot of big and great changes!

You bet!! 😄 

Don't hurry, take your time.

Naramsim commented:
Never heard about Webpack

sarbbottam commented:
To be honest I prefer a PR not more than 4 files changed or less that 200 LOC. But could not help in this one, they are so tightly coupled.

sarbbottam commented:
> Never heard about Webpack

It's great, and webpack 2.x is going to be awesome.

sarbbottam commented:
If you are still there, what is bithound for?

sarbbottam commented:
I would recommend to [clone the PR](https://github.com/sarbbottam/conf-files/blob/master/dot-commons/functions#L23-L35), as you review, I find it helpful when a PR involves, lot of changes.

I would suggest the following order of files for review:

Simple
- .eslintignore
- .gitignore

Deleted as they will be generated
- chrome/src/inject/inject.css
- chrome/src/inject/inject.js
- firefox/data/inject.css
- firefox/data/inject.js

Refactored
- src/change-color.js
- src/index.css
- src/index.js
- src/inject.js

Generate the inject.js
- webpack.config.js

Updated w.r.t. file and location
- chrome/manifest.json
- firefox/index.js

Dependencies and scripts
- package.json

Naramsim commented:
Ok, I got it all :smile: 
But there are two things that do not convince me. I will comment the code

Naramsim commented on package.json:
```diff
@@ -4,13 +4,10 @@
-    "build:watch": "watch \"npm run build\" ./src/",
+    "lint": "eslint .",
+    "prebuild": "npm run lint && rimraf chrome/src/ firefox/data/ build",
+    "build": "webpack -p",
```
Here Webpack compresses everything. When submitting a file to Opera and Mozilla store they must be able to read it or you can give them the instructions to get the same result with the compressed code.
I realized that Opera does not even try to build it and do not approves your extension. (I tried with vs1.5 and no response for a week)

sarbbottam replied:
That's easy, I have to remove `-p` production flag 😄

Naramsim commented on src/index.js:
```diff
@@ -0,0 +1,30 @@
+observer.observe(pjaxLoaderBar, config)
+
+// invoke changeColor function
+changeColor()
```
`changeColor()` must be invoked with `inject()`. If you don't when you click on repositories tab and back to overview, the squares are repainted in green colors

sarbbottam replied:
Ahh, make sense, so we can invoke `changeColor()` whenever `inject()` is being invoked.

sarbbottam commented:
@Naramsim could you take another look?
